contracts/utils/convex/ConvexYieldWrapper.sol

```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: BUSL-1.1
+
+pragma solidity 0.8.6;
+
+import '@yield-protocol/vault-interfaces/ICauldron.sol';
+import '@yield-protocol/vault-interfaces/DataTypes.sol';
+import './ConvexStakingWrapper.sol';
+
+/// @title Convex staking wrapper for Yield platform
+/// @notice Enables use of convex LP positions as collateral while still receiving rewards
+contract ConvexYieldWrapper is ConvexStakingWrapper {
+    using TransferHelper for IERC20;
+
+    /// @notice Mapping to keep track of the user & their vaults
+    mapping(address => bytes12[]) public vaults;
+
+    ICauldron public cauldron;
+
+    /// @notice Event called when a vault is added for a user
+    /// @param account The account for which vault is added
+    /// @param vaultId The vaultId to be added
+    event VaultAdded(address indexed account, bytes12 indexed vaultId);
+
+    /// @notice Event called when a vault is removed for a user
+    /// @param account The account for which vault is removed
+    /// @param vaultId The vaultId to be removed
+    event VaultRemoved(address indexed account, bytes12 indexed vaultId);
+
+    constructor(
+        address curveToken_,
+        address convexToken_,
+        address convexPool_,
+        uint256 poolId_,
+        address join_,
+        ICauldron cauldron_,
+        string memory name,
+        string memory symbol,
+        uint8 decimals
+    ) ConvexStakingWrapper(curveToken_, convexToken_, convexPool_, poolId_, join_, name, symbol, decimals) {
+        cauldron = cauldron_;
+    }
+
+    /// @notice Points the collateral vault to the join storing the wrappedConvex
+    /// @param join_ Join which will store the wrappedConvex of the user
+    function point(address join_) external auth {
+        collateralVault = join_;
+    }
+
+    /// @notice Adds a vault to the user's vault list
+    /// @param vaultId The id of the vault being added
+    function addVault(bytes12 vaultId) external {
+        address account = cauldron.vaults(vaultId).owner;
+        require(account != address(0), 'No owner for the vault');
+        bytes12[] storage vaults_ = vaults[account];
+        for (uint256 i = 0; i < vaults_.length; i++) {
+            require(vaults_[i] != vaultId, 'already added');
```

Comment:
Do we need to revert, or should we just return and do nothing if the cvault has already been added? Check with frontend engineering and make sure the implementation is consistent with `removeVault`